nil pointer startup visor on dmsghttp config

**Describe the bug**
Currently if STUN servers be empty, or all of them not available, then we get panic on visor startup due these logs:
```
[2022-06-30T03:18:23+04:30] WARN [stun_client]: Reply from 172.104.188.139:3478 server: UDP is blocked
[2022-06-30T03:18:23+04:30] WARN [stun_client]: All STUN servers offline
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xa07b21]

goroutine 28 [running]:
github.com/ccding/go-stun/stun.(*Host).IP(...)
	/home/mohammed/Projects/Skycoin/skywire/vendor/github.com/ccding/go-stun/stun/host.go:54
github.com/skycoin/skywire/pkg/visor.getPublicIP(0xc0001345a0, {0xc00023caa0, 0x4c})
	/home/mohammed/Projects/Skycoin/skywire/pkg/visor/init.go:1185 +0xe1
github.com/skycoin/skywire/pkg/visor.initAddressResolver({0x10be6a0, 0xc0001b0690}, 0xc0001345a0, 0xc0000381e0)
	/home/mohammed/Projects/Skycoin/skywire/pkg/visor/init.go:213 +0x97
github.com/skycoin/skywire/pkg/visor.withInitCtx.func1({0x10be6a0, 0xc0001b0690}, 0xc0001b0690)
	/home/mohammed/Projects/Skycoin/skywire/pkg/visor/init.go:1119 +0xb9
github.com/skycoin/skywire/pkg/visor/visorinit.(*Module).InitConcurrent(0x165c180, {0x10be6a0, 0xc0001b0690})
	/home/mohammed/Projects/Skycoin/skywire/pkg/visor/visorinit/module.go:136 +0x278
created by github.com/skycoin/skywire/pkg/visor/visorinit.(*Module).InitConcurrent
	/home/mohammed/Projects/Skycoin/skywire/pkg/visor/visorinit/module.go:115 +0x11c
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xa07b21]

goroutine 29 [running]:
github.com/ccding/go-stun/stun.(*Host).IP(...)
	/home/mohammed/Projects/Skycoin/skywire/vendor/github.com/ccding/go-stun/stun/host.go:54
github.com/skycoin/skywire/pkg/visor.getPublicIP(0xc0001345a0, {0xc00023cc30, 0x4c})
	/home/mohammed/Projects/Skycoin/skywire/pkg/visor/init.go:1185 +0xe1
github.com/skycoin/skywire/pkg/visor.initDiscovery({0x10be6a0, 0xc0001b0690}, 0xc0001345a0, 0xc0000381e0)
	/home/mohammed/Projects/Skycoin/skywire/pkg/visor/init.go:257 +0x1d5
github.com/skycoin/skywire/pkg/visor.withInitCtx.func1({0x10be6a0, 0xc0001b0690}, 0xc0001b0690)
	/home/mohammed/Projects/Skycoin/skywire/pkg/visor/init.go:1119 +0xb9
github.com/skycoin/skywire/pkg/visor/visorinit.(*Module).InitConcurrent(0x165c240, {0x10be6a0, 0xc0001b0690})
	/home/mohammed/Projects/Skycoin/skywire/pkg/visor/visorinit/module.go:136 +0x278
created by github.com/skycoin/skywire/pkg/visor/visorinit.(*Module).InitConcurrent
	/home/mohammed/Projects/Skycoin/skywire/pkg/visor/visorinit/module.go:115 +0x11c

```

--- a/pkg/visor/init.go
+++ b/pkg/visor/init.go
@@ -1199,7 +1199,7 @@ func getPublicIP(v *Visor, service string) (string, error) {
 }
 
 type ipAPI struct {
-	PublicIP string `json:"public_ip"`
+	PublicIP string `json:"ip_address"`
 }
 
 func getIP() (string, error) {

--- a/pkg/visor/init.go
+++ b/pkg/visor/init.go
@@ -2,8 +2,10 @@ package visor
 
 import (
 	"context"
+	"encoding/json"
 	"errors"
 	"fmt"
+	"io/ioutil"
 	"net"
 	"net/http"
 	"os"
@@ -210,7 +212,7 @@ func initAddressResolver(ctx context.Context, v *Visor, log *logging.Logger) err
 	}
 
 	// only needed for dmsghttp
-	pIP, err := getPublicIP(v, conf.AddressResolver)
+	pIP, err := getPublicIP(conf.AddressResolver)
 	if err != nil {
 		return err
 	}
@@ -254,7 +256,7 @@ func initDiscovery(ctx context.Context, v *Visor, log *logging.Logger) error {
 		factory.ServiceDisc = conf.ServiceDisc
 		factory.Client = httpC
 		// only needed for dmsghttp
-		pIP, err := getPublicIP(v, conf.ServiceDisc)
+		pIP, err := getPublicIP(conf.ServiceDisc)
 		if err != nil {
 			return err
 		}
@@ -790,7 +792,7 @@ func initUptimeTracker(ctx context.Context, v *Visor, log *logging.Logger) error
 		return err
 	}
 
-	pIP, err := getPublicIP(v, conf.Addr)
+	pIP, err := getPublicIP(conf.Addr)
 	if err != nil {
 		return err
 	}
@@ -992,7 +994,7 @@ func initPublicAutoconnect(ctx context.Context, v *Visor, log *logging.Logger) e
 		DiscAddr: serviceDisc,
 	}
 	// only needed for dmsghttp
-	pIP, err := getPublicIP(v, serviceDisc)
+	pIP, err := getPublicIP(serviceDisc)
 	if err != nil {
 		return err
 	}
@@ -1065,7 +1067,7 @@ func connectToTpDisc(ctx context.Context, v *Visor, log *logging.Logger) (transp
 	}
 
 	// only needed for dmsghttp
-	pIP, err := getPublicIP(v, conf.AddressResolver)
+	pIP, err := getPublicIP(conf.AddressResolver)
 	if err != nil {
 		return nil, err
 	}
@@ -1167,9 +1169,10 @@ func getHTTPClient(ctx context.Context, v *Visor, service string) (*http.Client,
 	return &http.Client{}, nil
 }
 
-func getPublicIP(v *Visor, service string) (pIP string, err error) {
+func getPublicIP(service string) (string, error) {
 	var serviceURL dmsgget.URL
-	err = serviceURL.Fill(service)
+	var pIP string
+	err := serviceURL.Fill(service)
 	// only get the IP from the stun client if the url is of dmsg
 	// else just send empty string as ip
 	if serviceURL.Scheme != "dmsg" {
@@ -1179,11 +1182,28 @@ func getPublicIP(v *Visor, service string) (pIP string, err error) {
 		return pIP, fmt.Errorf("provided URL is invalid: %w", err)
 	}
 
-	// Wait until stun client is ready
-	<-v.stunReady
-	if v.stunClient.PublicIP == nil {
-		return "", fmt.Errorf("all STUN servers are offline, use direct URL config instead dmsghttp")
-	}
-	pIP = v.stunClient.PublicIP.IP()
+	pIP = getip2()
 	return pIP, nil
+}
+
+type ipAPI struct {
+	Query string
+}
+
+func getip2() string {
+	req, err := http.Get("http://ip-api.com/json/")
+	if err != nil {
+		return err.Error()
+	}
+	defer req.Body.Close()
+
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return err.Error()
+	}
+
+	var ip ipAPI
+	json.Unmarshal(body, &ip)
+
+	return ip.Query
 }

--- a/pkg/visor/init.go
+++ b/pkg/visor/init.go
@@ -1173,7 +1173,7 @@ func getPublicIP(v *Visor, service string) (string, error) {
 	var serviceURL dmsgget.URL
 	var pIP string
 	err := serviceURL.Fill(service)
-	// only get the IP from the stun client if the url is of dmsg
+	// only get the IP if the url is of dmsg
 	// else just send empty string as ip
 	if serviceURL.Scheme != "dmsg" {
 		return pIP, nil

--- a/pkg/visor/init.go
+++ b/pkg/visor/init.go
@@ -1181,7 +1181,9 @@ func getPublicIP(v *Visor, service string) (pIP string, err error) {
 
 	// Wait until stun client is ready
 	<-v.stunReady
-
+	if v.stunClient.PublicIP == nil {
+		return "", fmt.Errorf("all STUN servers are offline, use direct URL config instead dmsghttp")
+	}
 	pIP = v.stunClient.PublicIP.IP()
 	return pIP, nil
 }

--- a/pkg/visor/init.go
+++ b/pkg/visor/init.go
@@ -1182,7 +1182,11 @@ func getPublicIP(service string) (string, error) {
 		return pIP, fmt.Errorf("provided URL is invalid: %w", err)
 	}
 
-	pIP = getip2()
+	pIP, err = getip2()
+	if err != nil {
+		return pIP, fmt.Errorf("cannot fetch public ip")
+	}
+
 	return pIP, nil
 }
 
@@ -1190,20 +1194,23 @@ type ipAPI struct {
 	Query string
 }
 
-func getip2() string {
+func getip2() (string, error) {
 	req, err := http.Get("http://ip-api.com/json/")
 	if err != nil {
-		return err.Error()
+		return "", err
 	}
-	defer req.Body.Close()
+	defer req.Body.Close() // nolint
 
 	body, err := ioutil.ReadAll(req.Body)
 	if err != nil {
-		return err.Error()
+		return "", err
 	}
 
 	var ip ipAPI
-	json.Unmarshal(body, &ip)
+	err = json.Unmarshal(body, &ip)
+	if err != nil {
+		return "", err
+	}
 
-	return ip.Query
+	return ip.Query, nil
 }

--- a/pkg/visor/init.go
+++ b/pkg/visor/init.go
@@ -212,7 +212,7 @@ func initAddressResolver(ctx context.Context, v *Visor, log *logging.Logger) err
 	}
 
 	// only needed for dmsghttp
-	pIP, err := getPublicIP(conf.AddressResolver)
+	pIP, err := getPublicIP(v, conf.AddressResolver)
 	if err != nil {
 		return err
 	}
@@ -256,7 +256,7 @@ func initDiscovery(ctx context.Context, v *Visor, log *logging.Logger) error {
 		factory.ServiceDisc = conf.ServiceDisc
 		factory.Client = httpC
 		// only needed for dmsghttp
-		pIP, err := getPublicIP(conf.ServiceDisc)
+		pIP, err := getPublicIP(v, conf.ServiceDisc)
 		if err != nil {
 			return err
 		}
@@ -792,7 +792,7 @@ func initUptimeTracker(ctx context.Context, v *Visor, log *logging.Logger) error
 		return err
 	}
 
-	pIP, err := getPublicIP(conf.Addr)
+	pIP, err := getPublicIP(v, conf.Addr)
 	if err != nil {
 		return err
 	}
@@ -994,7 +994,7 @@ func initPublicAutoconnect(ctx context.Context, v *Visor, log *logging.Logger) e
 		DiscAddr: serviceDisc,
 	}
 	// only needed for dmsghttp
-	pIP, err := getPublicIP(serviceDisc)
+	pIP, err := getPublicIP(v, serviceDisc)
 	if err != nil {
 		return err
 	}
@@ -1067,7 +1067,7 @@ func connectToTpDisc(ctx context.Context, v *Visor, log *logging.Logger) (transp
 	}
 
 	// only needed for dmsghttp
-	pIP, err := getPublicIP(conf.AddressResolver)
+	pIP, err := getPublicIP(v, conf.AddressResolver)
 	if err != nil {
 		return nil, err
 	}
@@ -1169,7 +1169,7 @@ func getHTTPClient(ctx context.Context, v *Visor, service string) (*http.Client,
 	return &http.Client{}, nil
 }
 
-func getPublicIP(service string) (string, error) {
+func getPublicIP(v *Visor, service string) (string, error) {
 	var serviceURL dmsgget.URL
 	var pIP string
 	err := serviceURL.Fill(service)
@@ -1182,20 +1182,28 @@ func getPublicIP(service string) (string, error) {
 		return pIP, fmt.Errorf("provided URL is invalid: %w", err)
 	}
 
-	pIP, err = getip2()
+	pIP, err = getIP()
 	if err != nil {
-		return pIP, fmt.Errorf("cannot fetch public ip")
+		<-v.stunReady
+		if v.stunClient.PublicIP != nil {
+			pIP = v.stunClient.PublicIP.IP()
+			return pIP, nil
+		}
+		err = fmt.Errorf("cannot fetch public ip")
+	}
+	if err != nil {
+		return pIP, err
 	}
 
 	return pIP, nil
 }
 
 type ipAPI struct {
-	Query string
+	PublicIP string `json:"public_ip"`
 }
 
-func getip2() (string, error) {
-	req, err := http.Get("http://ip-api.com/json/")
+func getIP() (string, error) {
+	req, err := http.Get("http://ip.skycoin.com")
 	if err != nil {
 		return "", err
 	}
@@ -1212,5 +1220,5 @@ func getip2() (string, error) {
 		return "", err
 	}
 
-	return ip.Query, nil
+	return ip.PublicIP, nil
 }

